### PR TITLE
WaveBot via Blueprint

### LINKED ISSUE #123: Add Single Body Device Model
As discussed in our last call, it makes a lot of sense to add the ability to model a single body device (probably in heave). This will help us encapsulate code that is shared between all models, and which is shared between similar classes, such as bodies in heave. 

A proposed case study for this model was [WaveBot](https://mhkdr.openei.org/submissions/151), but others candidates may also be viable.

### PR DESCRIPTION
**Pull request by ryancoe**
## Description
Porting #140 to the structure introduced by #144.

Fixes #123

## Checklist:

- [ ] Added the results of running the test suite (in pdf form)
- [ ] All new files contain the GPL header
- [ ] If `example.m` has been modified, the content / line numbers in 
  `docs\user\example.rst` are still valid or have been fixed

### CODE REVIEW

File: examples/WaveBot/WaveBot.m [hunk 1]
```diff
@@ -118,54 +127,39 @@
...
-        dw = mean(diff(S.w))*iSkip;   
-    end
-
-    % Get column vector S at same indicies as w (Removed 
-    % interpolation). 
-    s = S.S(iStart:iSkip:iEnd);
-
-    % TODO: is interp needed?
-    % s = interp1(S.w(:), S.S, w,'linear',0);
+    w = hydro.w(:);
```
Comment (ryancoe) [inline examples/WaveBot/WaveBot.m:130]: We should be much more structured about how we handle and check the properties of `w`. We should require on instantiation of a `Device` that it be

 1. a column vector
 2. fixed spacing
 3. monotonically increasing

File: examples/WaveBot/basic.m [hunk 2]
```diff
@@ -4,25 +4,33 @@
...
+
+figure
+plot(SS.w,SS.S)
+
+%%
 
 % make devices from blueprint. All arguments given as struct arrays
 % arrays with type and params field. 
 geomParams.type = 'scalar';
 geomParams.params = {1, S, 0.5};
```
Comment (ryancoe) [inline examples/WaveBot/basic.m:23]: Hard to know what's required here to be in the `geomParams` cell array

File: examples/WaveBot/basic.m [hunk 3]
```diff
@@ -4,25 +4,33 @@
...
 controlParams.type = 'CC';
 controlParams(2).type = 'P';
+controlParams(3).type = 'PS';
+controlParams(3).params = {10 1e9}; % {zmax, Fmax}
```
Comment (ryancoe) [inline examples/WaveBot/basic.m:27]: If you forget to specify `controlParams` for a `PS` controller, the error you receive is not super helpful.

File: examples/WaveBot/WaveBot.m [hunk 4]
```diff
@@ -118,54 +127,39 @@
...
-    else    
-        dw = mean(diff(S.w))*iSkip;   
-    end
-
-    % Get column vector S at same indicies as w (Removed 
-    % interpolation). 
-    s = S.S(iStart:iSkip:iEnd);
-
-    % TODO: is interp needed?
-    % s = interp1(S.w(:), S.S, w,'linear',0);
```
Comment (ryancoe) [inline examples/WaveBot/WaveBot.m:143]: I think all this skipping stuff has been or should be moved to the `SeaState` class, so I removed it from the `WaveBot` code.

File: examples/WaveBot/WaveBot.m [hunk 5]
```diff
@@ -118,54 +127,39 @@
...
+    w = hydro.w(:);
+    dw = w(2) - w(1); % TODO - make dw a property
+    
+    SS_struct = SS.struct;
+    SS_struct.S = interp1(SS_struct.w,SS_struct.S,hydro.w,...
+        'nearest',0); % TODO - allow user to set interp method
```
Comment (ryancoe) [inline examples/WaveBot/WaveBot.m:135]: We should allow the user to set this someone

File: examples/WaveBot/WaveBot.m [hunk 6]
```diff
@@ -81,17 +90,17 @@
...
+    staticModel.mass = hydro.Vo * hydro.rho;
 
     % Restoring
-    static.K = hydro.C(3,3) * hydro.g * hydro.rho;
+    staticModel.K = hydro.C(3,3) * hydro.g * hydro.rho;
 
 end
         
-function motion = getDynamicModel(static, hydro, S)
+function motion = getDynamicModel(staticModel, hydro, SS)
```
Comment (ryancoe) [inline examples/WaveBot/WaveBot.m:103]: I do not like the name `motion` -- to me, this sounds like a specific realization (i.e., results), not some fixed properties

File: toolbox/+WecOptTool/+types/Performance.m [hunk 7]
```diff
@@ -1,16 +1,16 @@
-classdef Performance < WecOptTool.base.Data
+classdef Performance < handle
```
Comment (ryancoe) [inline toolbox/+WecOptTool/+types/Performance.m:1]: I had a lot of trouble trying to do this with the inheritance @H0R5E had set up from the `Data` class. I really did try to work within that framework, but I eventually reverted to the dramatically changed version of the `Performance` class you see here.

File: examples/WaveBot/WaveBot.m [hunk 8]
```diff
@@ -81,17 +90,17 @@
...
 
     function result = interp_mass(hydro, dof1, dof2, w)
```
Comment (ryancoe) [inline examples/WaveBot/WaveBot.m:105]: `interp_mass` is a function that's only used once

File: examples/WaveBot/WaveBot.m [hunk 9]
```diff
@@ -178,50 +172,339 @@
...
+        
+        ph = ph_mat(:, ind_ph);
+        [phasePowMat(ind_ph), fRes(ind_ph), tRes(ind_ph)] = getPSPhasePower(motion, ph);
+        
+        
+        pos(:, ind_ph) = fRes(ind_ph).pos;
+        u(:, ind_ph) = fRes(ind_ph).vel;
+        Zpto(:, ind_ph) = fRes(ind_ph).Zpto;
+        Fpto(:, ind_ph) = fRes(ind_ph).u;
+        pow(:, ind_ph) = fRes(ind_ph).pow;
```
Comment (ryancoe) [inline examples/WaveBot/WaveBot.m:299]: Note that when looking at `PS` results, there are multiple realizations... I've decided to store them all and then make sure that `Performance` methods can handle this (i.e., by plotting the first realization only and my averaging the realizations for `summary`)

File: examples/WaveBot/WaveBot.m [hunk 10]
```diff
@@ -178,50 +172,339 @@
...
+        
+    % Fix random seed <- Do we want this???
+    rng(1);
+    
+    % Reformulate equations of motion
+    motion = getPSCoefficients(motion, delta_Zmax, delta_Fmax);
+    
+    % Add phase realizations
+    n_ph = 5;
+    ph_mat = [motion.ph, rand(length(motion.w), n_ph-1)];
```
Comment (ryancoe) [inline examples/WaveBot/WaveBot.m:287]: Note that I when making the `ph_mat` for PS, I have insert the `motion.ph` vector at the beginning. this allows for comparisons with other controllers since the phasing will reflect the `rng` state set by the user

File: examples/WaveBot/WaveBot.m [hunk 11]
```diff
@@ -118,54 +127,39 @@
...
 
     % intrinsic impedance
-    Zi = B + Bf + 1i * (w .* (static.mass + A) - static.K ./ w);
+    Zi = B + Bf + 1i * (w .* (staticModel.mass + A) - staticModel.K ./ w);
 
     % Excitation Forces
-    H = interp_ex(hydro, 3, w) * hydro.g * hydro.rho;
-    F0 = H .* eta_fd;
+    Hex = interp_ex(hydro, 3, w) * hydro.g * hydro.rho;
+    F0 = Hex .* eta_fd;
```
Comment (ryancoe) [inline examples/WaveBot/WaveBot.m:162]: I had a bug related to `F0` vs `Hex`: `F0 = Hex .* eta_fd`, so `Hex` is an intrinsic property of the WEC, whereas `F0` is specific to a sea state. I think everything is straight now, but we may want to think about ways to make this extra clear.

File: examples/WaveBot/WaveBot.m (hunk 1 above)
Comment (H0R5E) [reply examples/WaveBot/WaveBot.m:130]: Well, `w` is checked for the input of the seastate and in the RM3 example, that `w` is used rather than the one given to NEMOH. No reason why we can't check it the `Hydro` type as well though, if it makes to sense to be a variable used across all solvers (which I guess it does).

File: examples/WaveBot/basic.m (hunk 2 above)
Comment (H0R5E) [reply examples/WaveBot/basic.m:23]: I guess this is a decision the user makes when they define one of the geometryCallbacks. From the API:

> "each function should take a folder path as the first input argument, an arbitrary number of additional inputs and return a WecOptTool.types.Hydro object"

The "params" field contains the arbitrary inputs and I don't know how you would keep this interface flexible enough for people to define arbitrary additional inputs without something like this.

File: examples/WaveBot/basic.m (hunk 3 above)
Comment (H0R5E) [reply examples/WaveBot/basic.m:27]: Good spot, I can have a look at improving that.

File: examples/WaveBot/WaveBot.m (hunk 4 above)
Comment (H0R5E) [reply examples/WaveBot/WaveBot.m:143]: Yeah, SeaState is how the user can control this now.

File: examples/WaveBot/WaveBot.m (hunk 5 above)
Comment (H0R5E) [reply examples/WaveBot/WaveBot.m:135]: The approach to this would be to create another input to `getDynamicModel` and then use the `modelParams` input to [`makeDevices`](https://snl-waterpower.github.io/WecOptTool/master/user/api.html#+WecOptTool.Blueprint.makeDevices) to pass it. Because this is defined in the model, it's under the control of the user to add an additional parameter.

File: examples/WaveBot/basic.m [hunk 12]
```diff
@@ -4,25 +4,33 @@
...
-% w = 2*pi*linspace(0.05, 2, 50)';
-% S = jonswap(w,[Hm0, Tp, gamma],0);
-S = WecOptLib.tests.data.example8Spectra();
+Hm0 = 0.125;
+Tp = 2;
+gamma = 3.3;
+dw = 0.1;
+w = 2 + dw * (1:50)';
+S = jonswap(w,[Hm0, Tp, gamma],0);
+SS = WecOptTool.types("SeaState", S);
```
[the remark this answers is not in the document]
Comment (H0R5E) [reply examples/WaveBot/basic.m:13]: Unfortunately this is a problem of MATLAB's making. There in no way to build an array of handle objects (which the types are because of the use of dynamicprops) outside of a loop. Because you can't make them in the constructor, any array opertations you want to do on intialisation must also be done _after_ the array is created. So, if you were trying to make an array using the type directly it would go like:

```
for i = 1:length(S)
    result(i) = WecOptTool.types.SeaState.(S(i));
end

result.validateArray()
```

`validateArray` is a method I created so that any type can define array instantiation jobs. I wrote the `types` function to make this simpler to do.

[the remark this answers is not in the document]
Comment (H0R5E) [reply examples/WaveBot/basic.m:13]: I tried very hard to find another way.

File: examples/WaveBot/WaveBot.m (hunk 6 above)
Comment (H0R5E) [reply examples/WaveBot/WaveBot.m:103]: So, my logic is that the model defines the equations of motion (which I shortened to Motion here and for the type of the same name) which are then used to generate the performance of the device. That was the best I could do, but please feel free to change.

File: examples/WaveBot/WaveBot.m (hunk 8 above)
Comment (H0R5E) [reply examples/WaveBot/WaveBot.m:105]: It's just copy and paste from RM3. Defining these functions did reveal the bug I found, nonetheless.

Comment (H0R5E) [reply examples/WaveBot/WaveBot.m:105]: In that is shortens the code were it counts.

File: examples/WaveBot/WaveBot.m [hunk 13]
```diff
@@ -118,54 +127,39 @@
...
-    ph = rand(length(s), 1);
+    ph = rand(length(SS.S), 1);
 
     % Wave height in frequency domain
-    eta_fd = waveAmp .* exp(1i*ph);
+    eta_fd = waveAmp .* exp(1i * ph);
     eta_fd = eta_fd(:);
-    
     % radiation damping FRF
     B = interp_rad(hydro, 3, 3, w) * hydro.rho .* w;
```
[the remark this answers is not in the document]
Comment (H0R5E) [reply examples/WaveBot/WaveBot.m:149]: Same comment as for `interp_mass`.

File: toolbox/+WecOptTool/+types/Performance.m (hunk 7 above)
Comment (H0R5E) [reply toolbox/+WecOptTool/+types/Performance.m:1]: So, we differ in approach here. I have split the places where data is stored into 5 classes, whereas you are trying to get it all into one (super) class. I prefer my approach (naturally) because there is less chance for error from copying over the values that are already stored in other classes (which are accessible through the Device class). Your plot and summary functions can work on Devices if they need to draw from multiple sources, they can even be put into the device class, with a caveat...

All of the parameters that you define in the type classes are going to burden future users with defining them as well, so my advice would be to tread carefully and use them frugally. My concern is that if we demand variables to be defined that are not "universal", a user will need to change the framework code in order to not use them for their particular problem, which we don't want.

Let's talk about it during our meeting.

File: examples/WaveBot/WaveBot.m (hunk 9 above)
Comment (H0R5E) [reply examples/WaveBot/WaveBot.m:299]: Yeah, interesting. No reason why these can't be considered to be arrays in general and then work from there, as you have done. Different from RM3, which just took the mean, so we will need to update that example in unison.

File: examples/WaveBot/WaveBot.m (hunk 10 above)
Comment (H0R5E) [reply examples/WaveBot/WaveBot.m:287]: Something to change in RM3 as well?

File: examples/WaveBot/WaveBot.m (hunk 11 above)
Comment (H0R5E) [reply examples/WaveBot/WaveBot.m:162]: * This is going to sound harsh, I realise, but I'm just trying to define scope...

From a framework perspective this is the user's issue. Are we trying to teach the user how to understand this theory as well? Is that the plan for the docs?

File: toolbox/+WecOptTool/+types/Performance.m [hunk 14]
```diff
@@ -49,41 +49,225 @@
...
+            
+            for ii = 1:length(obj)
+                
+                for jj = 1:size(obj(ii).ph,2) % for each phase in PS cases
+                
+                    tmp.pow_avg(ii,jj) = sum(real(obj(ii).pow(:,jj)));
+                    
+                    pow_t = getTimeRes(obj(ii), 'pow', t, jj);
+                    tmp.pow_max(ii, jj) = max(abs(pow_t));
+                    tmp.pow_thd(ii, jj) = thd(pow_t);
```
Comment (ssolson) [inline toolbox/+WecOptTool/+types/Performance.m:193]: `thd` requires the signal processing toolbox. Do we want to write a function for this?

Comment (ssolson) [reply toolbox/+WecOptTool/+types/Performance.m:193]: MHKiT just added a `total_harmonic_current_distortion` function based on IEC TS 62600-30:2018 ED1 or IEC TS 61000-4-7:2008 ED2 (I'm not sure which one). Anyways this could be a good reference for starting.

Comment (ryancoe) [reply toolbox/+WecOptTool/+types/Performance.m:193]: I'd rather maybe do a `try-catch`...

File: examples/WaveBot/WaveBot.m (hunk 1 above)
Comment (ryancoe) [reply examples/WaveBot/WaveBot.m:130]: There are two distinct frequency vectors

1. Seastate(s) - currently checks the frequency vector
2. BEM/frequency domain calcs./pseudo-spectral - no formal storage of `w`

**Solution (to be addressed in a separate issue): add an attribute `w` to [`WecOptTool.types.Hydro`](https://github.com/SNL-WaterPower/WecOptTool/blob/master/toolbox/+WecOptTool/+types/Hydro.m) with the above required properties**

File: examples/WaveBot/basic.m (hunk 2 above)
Comment (ryancoe) [reply examples/WaveBot/basic.m:23]: **solution: add docstrings to device classes and inline code the examples**

File: examples/WaveBot/basic.m (hunk 3 above)
Comment (ryancoe) [reply examples/WaveBot/basic.m:27]: @H0R5E will address this in this PR

File: examples/WaveBot/WaveBot.m (hunk 5 above)
Comment (ryancoe) [reply examples/WaveBot/WaveBot.m:135]: @H0R5E will address in this PR

File: examples/WaveBot/basic.m (hunk 12 above)
[the remark this answers is not in the document]
Comment (ryancoe) [reply examples/WaveBot/basic.m:13]: Understood, we'll come back to this later if anyone has more ideas

File: examples/WaveBot/WaveBot.m (hunk 6 above)
Comment (ryancoe) [reply examples/WaveBot/WaveBot.m:103]: will change `motion` to `motionModel`

File: toolbox/+WecOptTool/+types/Performance.m (hunk 7 above)
Comment (ryancoe) [reply toolbox/+WecOptTool/+types/Performance.m:1]: Good discussion on this... some key points:

1. Don't want to create onerous requirements (i.e., some results may not be created for a certain device, don't want to require them)
2. Adding complexity is bad... I found this really hard to understand. My opinion is that we should sometimes make the code less capable (and maybe less elegant) but more transparent.

File: toolbox/+WecOptTool/+types/Performance.m (hunk 14 above)
Comment (ryancoe) [reply toolbox/+WecOptTool/+types/Performance.m:193]: let me try to add a `try-catch`

File: examples/WaveBot/WaveBot.m (hunk 11 above)
Comment (ryancoe) [reply examples/WaveBot/WaveBot.m:162]: I will look back through the code and see if there's anything that can be done to make this more clear (could've just been a silly mistake on my part)

**Comment by H0R5E** [comment]
OK, so I've put an example of the plots and summary table you added to the Performance class as functions that work on devices, in the waveBot example folder. You can follow it through from the `WaveBot_caseA_Mat.m` file.

Having looked at the waveBot code, compared it to the RM3 code, and from listening to our conversation last Thursday, I really think that dropping the concept of a framework for something like library supported examples is going to work better for the project at this stage.

In particular, I feel like it is an unwanted burden to fit new examples into the same structure each time or to update old examples to new code. For example, all of the analysis code for waveBot only considers a single sea-state problem and the outputs of the controllers differ from the RM3 case significantly (they are complex, for instance). It's clear there is still a desire to tweak the functionality particular to a example rather to the framework as a whole, so that all of the examples do not need to be changed at once.

So, in conclusion, I think we should get rid of all the classes (perhaps baring the SeaState class) and try and unroll them into two (super) functions that output structs and can be edited as required depending on the needs of the example at hand. In that case, the user can build the analysis that is most appropriate to them at the time (like not having to worry about dealing with multiple sea states if they are not going to use them or using the functionality of different toolboxes). 

I don't think it would take me very long to do this, if you want me to have a crack at it?

**Comment by H0R5E** [comment]
As as example of what I am proposing see #161. I'll finish off the optimisation example tomorrow.

**Comment by H0R5E** [comment]
Yeah, so having got both examples working in #161, I'm strongly recommending moving over to the concept represented there (i.e. structured examples with a library of independent tools to help with the problems). Much as I have pushed a top down approach, I think it's evidently suffocating development of algorithms trying to put a structure over the entire problem at this stage. The "structured examples" approach in #161 will give a lot more flexibility in how the problems are built, but with the aid of "soft" framework for how they might go together. I think we can come back to trying to put a more formal data structure on the problem in the future (although it will be a task to get every example made to conform at that point) once it's clearer what all of the stages are going to produce. Parts of the structure might be transformed anyway if we turn to more complex tools for equation generation.

So, although binning 2000 lines of code and docs seems like a bit of a waste, I think it has been a useful exercise to "birth" the the two main functions in #161 and as it's stored in the history of the trunk, it can always be recovered at a later date should standardising the approach become more important.

**Comment by H0R5E** [comment]
This is now incorporated into #161, so can be closed.

**Comment by ryancoe** [comment]
now under #161
